### DISCUSSION TITLE
Well Done, and a suggestion ....

### DISCUSSION THREAD

**Discussion by chuckm2000** [kg4vdk/arcos-linux-modules#11] [General Discussion]
I have just done a test run of arcOS v21.3 on a laptop and I'm very impressed.  I have been a Linux/BSD/macOS/Windows user for many years, and this is one of the best implementations layered on top of Linux Mint I have ever seen.  Well done!  I can see why a non-computer-geek user could easily use this distro with little fuss getting it running.

A suggestion I'd love to see implemented ... an easy way to permanently install arcOS on the laptop's SSD without having to remove it to 'dd' the ISO directly to it.  Performance should be excellent on even a modest i5 CPU.

Many thanks for this, 73!
Chuck. VA7UL VA7ULD
va7ul@va7ul.ca

**Comment by kg4vdk** [comment]
Hey Chuck,

Thanks for the kind words and for testing it out! For installing to an internal SSD (which I do on my machines with less than plentiful USB ports), my method is to boot from an arcOS USB, with a copy of the ISO on board. Then, I use the "Disks" application to "Restore Disk Image" using the SSD as the target and the ISO as the source.

Heads up...this can make upgrading the ISO on the SSD difficult in the future, as the ISO-9660 signature is written to the SSD. This makes attempting to boot from a USB drive hard, because the system still sees a "bootable CDROM" signature on the SSD. My solution is to boot from the internal arcOS drive, and use the `fdisk` or `wipefs` utility to wipe the boot partition. Then, booting from the USB works as expected.

If I can come up with a clean way to make this easier for normal users, I'll certainly include it. Thanks for the input! I'm going to convert this issue to a discussion, so maybe others can weigh in.

**Reply by chuckm2000** [reply]
Thanks Mike!   I've been thinking about this a bit more ... it would not be a trivial task, but allowing permanent installs on a computer's internal drive without having to use a flash stick for subsequent updates is an ideal environment.  But that means coming up with a process that simply allows your updates to come from from a repo somewhere ... and setting up a repository isn't exactly trivial.  Of course that's how mainline distros do it, but would you have enough followers/users to make the significant effort worthwhile?

Just a further thought :-)   In the mean time I'll start messing with arcOS as spare time permits ... lots of fun to discover .  I strongly suspect arcOS would make the ideal foundation for a portable computer for emergency comms and field day use.

73,  Chuck

**Reply by kx8u** [reply]
Mike,

I too just installed Arcos and really like it. It seems to work great on an Evolve III laptop. Good to know it support cheap, slow equipment. I would like to also request, as Chuck did, a way to install to the local hard drive. The USB stick for field use is asking for disaster plus I have only 2 USB ports. Using a powered USB hub in the field is a non-starter for me and the GPS won't power up on one. Plus I need a mouse and Digirig. Eliminating the OS on a USB would be a great help. I have seen Linux distros in the past that allowed you to install from the live disk if you wanted. I know that is built not Linux Mint as I have used it before. Maybe you can add that to your distro? Just a thought. Thanks for all the work this, really clean and good job.

73,

David
KX8U

**Reply by kg4vdk** [reply]
Hey David,

Glad you like it, and that it supports your hardware. I don't have an Evolve (or the Wowee) to test with, so it's nice to know that the inclusion of the 6.5 kernel handled the formerly built-from-source drivers which were required for that hardware. I'm interested to know which GPS unit you are using that doesn't work correctly without a powered USB hub. There are a couple of affordable ones listed on the Documentation page of the website that I use regularly with an unpowered Anker brand USB hub. I'll reiterate that arcOS CAN, indeed, be installed onto an SSD internally...with the caveats listed in the response to Chuck, above. 

The way arcOS is set up, the OS itself is immutable, though features and functionality can be added through the use of modules. However, some things that might need to be fixed will require a rebuild of the ISO image. Maintaining the "unbreakable" core OS is paramount to this project being what it set out to be. If more "advanced" users want to write the image to an SSD, I'm all for it. The commands needed to upgrade in the future are simple, and I'll likely document them more thoroughly as I get the time. Thanks for the feedback, and for engaging here! It means a lot!

73 de KG4VDK
Mike

**Comment by kx8u** [comment]
Mike,Again thanks so much for you work to improve our Ham Radio service. I to backtrack on my GPS comment. It to almost 20 minutes for it to finally flash the first light so I thought it was dead. I’m using a GPS/GLONASS U-blox 7. I have had them for years. Guess it just came up way slow this time but it did start working. Your OS build works great on the Evolve. Since you don’t have one I will be happy to test for you if you want me too. Not sure if that is needed but feel free to reach out if you need the help. Yes please publish any detailed instructions on installing to the hard drive. I am not the good with Linux so the help will be great. You could even do a video that details how to do it. This could be similar to KM4ACK does install videos. Maybe he could even work with you on one. I do like your “can’t break it” design however. DavidKX8UOn Aug 7, 2024, at 17:54, KG4VDK ***@***.***> wrote:﻿
Hey David,
Glad you like it, and that it supports your hardware. I don't have an Evolve (or the Wowee) to test with, so it's nice to know that the inclusion of the 6.5 kernel handled the formerly built-from-source drivers which were required for that hardware. I'm interested to know which GPS unit you are using that doesn't work correctly without a powered USB hub. There are a couple of affordable ones listed on the Documentation page of the website that I use regularly with an unpowered Anker brand USB hub. I'll reiterate that arcOS CAN, indeed, be installed onto an SSD internally...with the caveats listed in the response to Chuck, above.
The way arcOS is set up, the OS itself is immutable, though features and functionality can be added through the use of modules. However, some things that might need to be fixed will require a rebuild of the ISO image. Maintaining the "unbreakable" core OS is paramount to this project being what it set out to be. If more "advanced" users want to write the image to an SSD, I'm all for it. The commands needed to upgrade in the future are simple, and I'll likely document them more thoroughly as I get the time. Thanks for the feedback, and for engaging here! It means a lot!
73 de KG4VDK
Mike

**Reply by kg4vdk** [reply]
Check out why that GPS might take a while to get a fix sometimes. Here's a pretty good explanation:

https://en.wikipedia.org/wiki/Time_to_first_fix

Keep truckin' with the Evolve, and let me know if you run across any issues!

**Comment by chuckm2000** [comment]
Some things I have noted in early testing .... 

Firefox tends to get stuck a lot and can occasionally just freeze.  I'm on a fast network so speed is not an issue.  It's the same on Ethernet or on WiFi.  I would not be concerned about this until I can test with the OS installed on the laptop's SSD.  I think Firefox likes to write to the flash stick a lot which probably explains the issue.

I am testing USB GPS dongles, and currently have an inexpensive G-Mouse brand using the VK-162 chip.  It sees a lot of satellites but is not very sensitive, so SNR's are not good.  Last evening I ordered a better (more expensive) Taiwanese dongle at Amazon which arrives later today, a GlobalSat BU-353N5 which should be much better.  One good thing is that arcOS immediately recognizes the dongle.

I don't know much about Winlink yet, never having used it, so I will learn how to get it up and running.  I have a Kantronics 3 Plus packet modem which is immediately recognized by Linux if I look at 'dmesg', but I'm not sure how to get Winlink (or PAT??) to connect with it.   Something to explore.  I do need to register with Winlink to get started, of course.

I'll keep you posted as I gradually navigate through arcOS, learning all sorts of new stuff .. Hihi.   I have had it running 24/7 for a few days with no issues regarding stability.  Your choice of Mint as the platform was perfect, and avoiding Cinnamon's Wayland server in favour of traditional X11 solves a lot of potential problems (xgps for example, doesn't like Wayland). 

So far, so good!  73,  Chuck

**Reply by kg4vdk** [reply]
For the Firefox issue, yeah, it does like to make lots of writes. Is this using a USB 3.0 drive? I did some testing EARLY in this project with USB 2.0 drives I had laying around. They are terribly slow. I haven't noticed this issue with Firefox hanging on my end, and I often have 3-5 tabs open at a time. 

I use the GlobalSat BU-353 in my truck, and it has been great, though like you said, it's a pricier one.

While you may certainly want to try getting the KPC3+ working, if you have a Digirig you won't need the KPC. Direwolf is the software TNC. When you run Pat, by default it is set up for telnet (internet), ax25 (packet), and ARDOP. If neither Direwolf or ARDOP is running, Pat can use telnet. If Direwolf/ARDOP is running, Pat is configured for use with them already.

If you aren't already familiar with setting up a Winlink account, here's the blurb from the https://winlink.org/user page:

Follow the program's instructions to configure it for your callsign and use (handled by arcOS). Connect with the system (sending a message via Telnet is okay) to begin creating your account. Do not use a password on your first connection. Your account's radio email address is YOURCALL@winlink.org. A message containing your password will be posted by the system to your account. Retrieve it with a second connection. After retrieving your password, secure login will now be enforced by the CMS, so be sure to set your password in your client program before your next connection try (or at least write it down, so you can enter it manually)!

**Comment by chuckm2000** [comment]
Wow, that is great info.  Sounds like I should simply use the software 
TNC, which I suspect is more versatile than the KPC which limits me to 
1200 Baud.  My KPC has an RS232 connection directly into an Alinco 2M 
radio, so that would still be useful on VHF although it has not yet been 
properly set up.  I should also look at Digirig and SignalLink devices.

All part of the learning curve!   I’ll get my digital call (VA7ULD) 
registered with Winlink asap.

Thanks muchly!  Chuck.  VA7UL


------ Original Message ------
From "KG4VDK" ***@***.***>
To "kg4vdk/arcos-linux-modules" ***@***.***>
Cc "chuckm2000" ***@***.***>; "Author" ***@***.***>
Date 2024-08-09 11:23:11 AM
Subject Re: [kg4vdk/arcos-linux-modules] Well Done, and a suggestion 
.... (Discussion #11)

**Comment by chuckm2000** [comment]
More discoveries, none of which are critical ....
One problem that a live Flash stick can't always fix ... entries in the hosts file and gpsd config file don't persist, and I have to re-enter my WiFi password every time I reboot, but those are picky points that will likely be solved once we can install on the computer's disk.

And ....
I certainly agree with KX8U, and totally approve of the immutable core system.  Ever since I found Apple's switch to an immutable system for macOS I have seen security take a big leap forward, so I'm pleased to see the same thing with Mint. 

A very useful feature is the system status report at the lower right of the screen ... important stuff at a glance.  I really like that.

73!   Chuck

**Comment by kg4vdk** [comment]
> entries in the hosts file and gpsd config file don't persist

Since these files are part of the "immutable" side of things, this is expected. To make it persist, you can use a USER module/script. Just make a directory named appropriately for your use, and place the modified version of /etc/default/gpsd and /etc/hosts there, and have the script simply copy them into place and then restart the relevant service. Then, when the Station Setup runs through the module directories, your changes will be put into effect.

> I have to re-enter my WiFi password every time I reboot

This SHOULD be handled by the "SAVE WIFI CONNECTION" menu item under "arcOS Tools" in the main menu. Once you've connected to a network, just run the save script, and that connection information will be saved and available on future boots.

> picky points that will likely be solved once we can install on the computer's disk

I don't think they are picky, but they won't be solved by writing the ISO to the internal disk. That won't make anything more persistent. It will only free up the USB port. The live operating system is integral to being able to provide support. The moment it becomes a "full install", all bets are off. There's no telling what manner of things people can do that affect more than they expect, and troubleshooting without a common starting point is often fruitless. 

Mike

**Reply by chuckm2000** [reply]
Good points, Mike!  I didn't think about how users could dig a big hole and pull the dirt in behind them if they get to mess with an installed system.   So I can see a big advantage staying with a flash stick base, if only to keep the support request noise manageable.  Your suggestion of a script is a good one ... I enjoy shell scripting so I'll do that.

Too bad my laptop won't boot from USB-3 even though it has a port .... a common problem with older laptops having no USB-3 driver in the BIOS.  I have to boot my USB-3 flash stick from a USB-2 port.  Oh well ......  :-)

**Comment by nx2i** [comment]
I have been experimenting with arcOS a little and I have some comments and questions relating to the experience.
Firstly, thanks for creating this, Mike. I think the idea is great and will be helpful to a great many hams. I have significant user experience with Linux, going all the way back to Solaris....pretty much as a user and admin; not a programmer. here's some comments and a question:
1) startup and shutdown are really pretty good (speedy). I have tried arcOS on 2 different laptops, an ancient Dell D630 and a more recent Dell D7390-2n1. I will soon experiment with both a gps unit and digirig soundcard as I have both on hand.
2) my USB flash runs really warm. The temp is probably not excessive., but I have never before noticed a flash drive feeling hot to the touch after removing the drive. It's a SanDisk Ultra 3.0 32GB, BTW.
3) How portable is the system meant to be? I first used the flash drive with arcOS on the D7390. The system is wifi only. All the initial configuration went well. When I tried to use the same flash drive on the D630, I can only get the wired internet configuration to work. I have tried a few times to save the wireless config, but it only seems to want to use the ethernet wired interface. The wired interface is what it saw on the first boot. Am I doing something incorrectly?
73, John, NX2I

**Reply by kg4vdk** [reply]
Hey John,

Thanks for the feedback!

My USB drives are also warm, but I've had no issues so far with quality drives. Some of the metal bodied drives probably dissipate heat a bit better.

With regard to the WiFi issue...I understand this to mean that the D7390 works as expected. Can you clarify about the D630? Are you able to connect to WiFi, but just not save it and have it work on a reboot? Or, does the D630 have issues connecting in the first place?

How it is "supposed" to work:
Boot from USB on Machine A, connect to WiFi, save WiFi config. Machine A should reconnect to the WiFi automatically on reboots.
This should also allow Machine B to connect to the same WiFi if booted from the same USB.

**Comment by nx2i** [comment]
I will perform some more tests. On Machine A, I can only boot with wifi as the interface; there is not an ethernet interface. Everything I have tested works great. If I reboot on A all is well. On machine B, there are both ethernet and wifi interfaces. The first boot on B made an ethernet connection seamlessly. Subsequent attempts to enable (not necessarily save) the wifi config have failed. This is true even though the ss-id and password for wifi are the same for both A and B.
I'll do more testing tomorrow.
John

**Reply by kg4vdk** [reply]
Ok, let's put the "working" D7390 over to the side and focus on the D630. Remove any saved wifi configs from the QRV/CALLSIGN/SAVED/WIFI directory. With this "clean" setup, can the D630 connect to the wifi as expected? If not, please use the "System Info" tool from the main menu and use the "Copy to clipboard" option to place the information in a .txt file, and upload the file here. I'm curious abut the wifi hardware on-board.

Thanks for your continued efforts and testing!

**Reply by nx2i** [reply]
I ran the D630 machine for awhile with the ethernet interface. I had both direwolf and YAAC running well.
I remove the wifi config file and rebooted but this time with the ethernet cable disconnected. I could not get the wifi to run.
The system info file is attached.
[D630-system.txt](https://github.com/user-attachments/files/16620457/D630-system.txt)

**Reply by nx2i** [reply]
BTW, the "normal" OS on the D630 is Linux Mint 21.3 (mate) and the wifi interface works just fine.

**Reply by kg4vdk** [reply]
After some digging, here's what I have found:

It appears that the live ISO does not enable the proprietary drivers for the BCM4312. A solution was offered a few years back by one of the moderators over on the LM forums regarding this issue, alluding to it working on the live ISO without a reboot. You might try this to see if it works for you...

```
sudo apt install firmware-b43-installer
sudo modprobe -r b43
sudo modprobe -r ssb
sudo modprobe b43
```
https://forums.linuxmint.com/viewtopic.php?p=1855702#p1855702

Let me know if this helps make progress, and I can help you make the changes persistent by putting the necessaries in a USER module.

One more thing to check (and I doubt it's the case), but what's the output of `rfkill list`? I'd hate to overlook the possibility of the WiFi hardware switch being in the off position. :-)

**Reply by nx2i** [reply]
GM Mike,
you are correct on all counts. Firstly, rfkill was not an issue. Prior to executing the commands listed above, netmanager would show nothing in the Device dropdown list under the wifi tab. After executing the above, I could select wlan0 in the dropdown and the  wifi works well. Also as you would expect the changes are not persistent.
Thanks for making this and thanks for the help.
73, John

**Reply by kg4vdk** [reply]
Sounds a lot like progress to me! Good work!

**Reply by nx2i** [reply]
Mike,
when you have a chance I would like to make a user module to incorporate the changes necessary to enable wifi on the d630.
Knowing this would likely allow me to help some of the local groups. No big rush. I have looked a the template module in the Community Directory.
73, John

**Reply by kg4vdk** [reply]
John,

Attached is the module I have come up with. Let me know if this method works for you!

To use this module, extract the `.tar.gz` file, and place the BROADCOM-B43 directory inside the USER module directory. Then, move the 00_BROADCOM-B43.sh script up one level from it's current location (inside the module directory), so that the 00_BROADCOM-B43.sh script resides in the USER directory. Here's what it does:

First, it installs the `b43-fwcutter` package (without requiring internet access, as the .deb file is included in the `packages` directory.

Second, instead of installing/using the `firmware-b43-installer` package, the firmware files are included inside the `broadcom-wl-6.30.163.46.tar.bz2` archive. Once extracted, we have the `broadcom...apsta.o` file available. 

Third, we use the `b43-fwcutter` to extract the firmware files into their correct location in `/usr/lib/firmware`. We clean up by deleting the no-longer-needed `broadcom...apsta.o` file. We, also set the correct permissions on the new `/usr/lib/firmware/b43` directory.

Last, we remove conflicting modules, and reload the b43 module.

LINK REMOVED

**Comment by nx2i** [comment]
Mike,
I  placed the BROADCOM-B43 directory in the USER directory:
user@arcOS:/media/user/ARCOS-DATA/QRV/NX2I/arcos-linux-modules/USER$ ls
00_BROADCOM-B43.sh  BROADCOM-B43  README
and copied the script up one level. I rebooted the machine, but did not get the desired result.

I tried running the script from the command line. It ran and produced 
 a new file in the directory:
user@arcOS:/media/user/ARCOS-DATA/QRV/NX2I/arcos-linux-modules/USER$ ls 
00_BROADCOM-B43.sh  BROADCOM-B43  README  broadcom-wl-6.30.163.46.wl_apsta.o

The log file shows:

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
b43-fwcutter is already the newest version (1:019-7build2).
0 upgraded, 0 newly installed, 0 to remove and 297 not upgraded.
broadcom-wl-6.30.163.46.wl_apsta.o
Cannot open input file /media/user/ARCOS-DATA/QRV/NX2I/arcos-linux-modules/USER/BROADCOM-B43/broadcom-wl-6.30.163.46.wl_apsta.o
rm: cannot remove '/media/user/ARCOS-DATA/QRV/NX2I/arcos-linux-modules/USER/BROADCOM-B43/broadcom-wl-6.30.163.46.wl_apsta.o': No such file or directory
chmod: cannot access '/usr/lib/firmware/b43': No such file or directory

BTW there are 2 USER directories. I assume I picked the correct one...
John

**Reply by nx2i** [reply]
is it looking in the wrong directory?

**Reply by kg4vdk** [reply]
I did some cleanup and fixing...

The issue appears to have been that the `tar`command was extracting the `...apsta.o` file to the wrong directory.

Attaching a "fixed" version for testing:

[BROADCOM-B43_TESTING.tar.gz](https://github.com/user-attachments/files/16679288/BROADCOM-B43_TESTING.tar.gz)

I am unclear what you mean by "there are 2 USER directories." There should only be 1 "USER" directory found at `/media/user/ARCOS-DATA/QRV/$CALLSIGN/arcos-linux-modules/USER`. Can you provide more details about the second directory?

**Reply by nx2i** [reply]
Mike, I have been working on updating my setup to "India" from "Golf" The script you wrote in the tar file mentioned here, no longer works for me on my laptop with the broadcom wireless issue. I get the error message that the script fails.
Interestingly, if I follow the procedure you mentioned above:
sudo apt install firmware-b43-installer
sudo modprobe -r b43
sudo modprobe -r ssb
sudo modprobe b43,
this works,. But of course it is not a "persistent" change. Any suggestions?
73, NX2I

**Reply by kg4vdk** [reply]
This appears to be related to the path issue we discussed on https://github.com/kg4vdk/arcos-linux-modules/discussions/40. Line 17 rears its head again! :-)

**Comment by nx2i** [comment]
Mike,
the modified module works well. Thanks.
I am not sure how I ended up with 2 USER directories. The one I have since
deleted was located at :
/media/user/ARCOS-DATA/QRV/NX2I/USER. Removing it has not created any
problems. I may have accidently
created this when moving the BROADCOM-B43 module from Downloads.
73, John

**Reply by kg4vdk** [reply]
Glad to hear it! I think we've all gone a bit wild and CTL+V'd when we didn't mean to. Lol!

**Comment by nx2i** [comment]
Roger that..

**Comment by nx2i** [comment]
I have fixed the issue. Just changed yhe arcos-data directory. Tnx & 73.
Nx2i
